cpp/src/parquet/statistics_test.cc

```diff
@@ -874,160 +875,129 @@ TEST_F(TestStatisticsSortOrderFLBA, UnknownSortOrder) {
+  stats->Update(values.data(), values.size(), 0);
+  ASSERT_FALSE(stats->HasMinMax());
+}
+
+template <typename Stats, typename Array>
+void AssertUnsetMinMax(Stats stats, const Array& values, const uint8_t* valid_bitmap) {
+  auto n_values = values.size();
+  auto null_count = ::arrow::internal::CountSetBits(valid_bitmap, n_values, 0);
+  auto non_null_count = n_values - null_count;
+  stats->UpdateSpaced(values.data(), valid_bitmap, 0, non_null_count, null_count);
+  ASSERT_FALSE(stats->HasMinMax());
+}
+
+// PARQUET-1225: Float NaN values may lead to incorrect min-max
+template <typename ParquetType>
+void CheckNaNs() {
+  using T = typename ParquetType::c_type;
+
+  constexpr int kNumValues = 8;
+  NodePtr node = PrimitiveNode::Make("f", Repetition::OPTIONAL, ParquetType::type_num);
   ColumnDescriptor descr(node, 1, 1);
-  float values[NUM_VALUES] = {std::nanf(""), -4.0f, -3.0f, -2.0f, -1.0f,
-                              std::nanf(""), 1.0f,  2.0f,  3.0f,  std::nanf("")};
-  float nan_values[NUM_VALUES];
-  for (int i = 0; i < NUM_VALUES; i++) {
-    nan_values[i] = std::nanf("");
-  }
-  std::vector<uint8_t> valid_bits(BitUtil::BytesForBits(NUM_VALUES) + 1, 255);
+
+  constexpr T nan = std::numeric_limits<T>::quiet_NaN();
+  constexpr T min = -4.0f;
+  constexpr T max = 3.0f;
+
+  std::array<T, kNumValues> all_nans{nan, nan, nan, nan, nan, nan, nan, nan};
+  std::array<T, kNumValues> some_nans{nan, max, -3.0f, -1.0f, nan, 2.0f, min, nan};
+  uint8_t valid_bitmap = 0x7F;  // 0b01111111
+  // NaNs excluded
+  uint8_t valid_bitmap_no_nans = 0x6E;  // 0b01101110
 
   // Test values
-  auto nan_stats = MakeStatistics<FloatType>(&descr);
-  nan_stats->UpdateSpaced(&values[0], valid_bits.data(), 0, NUM_VALUES, 0);
-  float min = nan_stats->min();
-  float max = nan_stats->max();
-  ASSERT_EQ(min, -4.0f);
-  ASSERT_EQ(max, 3.0f);
-
-  // Test all NaNs
-  auto all_nan_stats = MakeStatistics<FloatType>(&descr);
-  all_nan_stats->UpdateSpaced(&nan_values[0], valid_bits.data(), 0, NUM_VALUES, 0);
-  min = all_nan_stats->min();
-  max = all_nan_stats->max();
-  ASSERT_TRUE(std::isnan(min));
-  ASSERT_TRUE(std::isnan(max));
-
-  // Test values followed by all NaNs
-  nan_stats->UpdateSpaced(&nan_values[0], valid_bits.data(), 0, NUM_VALUES, 0);
-  min = nan_stats->min();
-  max = nan_stats->max();
-  ASSERT_EQ(min, -4.0f);
-  ASSERT_EQ(max, 3.0f);
-
-  // Test all NaNs followed by values
-  all_nan_stats->UpdateSpaced(&values[0], valid_bits.data(), 0, NUM_VALUES, 0);
-  min = all_nan_stats->min();
-  max = all_nan_stats->max();
-  ASSERT_EQ(min, -4.0f);
-  ASSERT_EQ(max, 3.0f);
-
-  // Test values followed by all NaNs followed by values
-  nan_stats->UpdateSpaced(&values[0], valid_bits.data(), 0, NUM_VALUES, 0);
-  min = nan_stats->min();
-  max = nan_stats->max();
-  ASSERT_EQ(min, -4.0f);
-  ASSERT_EQ(max, 3.0f);
+  auto some_nan_stats = MakeStatistics<ParquetType>(&descr);
+  // Ingesting a mix of NaNs and non-NaNs should not yield valid min max.
```

Comment:
Why? Is this what the Parquet spec says? It doesn't sound very useful.

Reply:
I can't find anything in the specification except that if NaN is found in the serialized statistics then the statistics should not be used. Is parquet-mr refusing to compute statistics when there is any NaN?

I looked at the Impala sources and it seems that Impala will write NaN statistics. Failing to write statistics when there are NaN values at least _is not harmful_.

Reply:
It's not according to spec, but [this](https://github.com/apache/parquet-mr/blob/dc61e510126aaa1a95a46fe39bf1529f394147e9/parquet-column/src/main/java/org/apache/parquet/column/statistics/Statistics.java#L86-L102). It's not refusing to compute statistics.

Reply:
I'd say it depends what other implementations do with statistics on reading. Do they assume that non NaNs are contained if the statistics are present?

Reply:
This is specified in the [spec](https://github.com/apache/parquet-format/blob/master/src/main/thrift/parquet.thrift#L881-L889).

Reply:
To summarize:

* There may be NaN values even if min/max are set
* If the min or max are NaN, then the statistics should be ignored for predicate evaluation purposes (data page or row group skipping)

I think we should compute the min/max value excluding NaN values if at all possible. Unless we believe that this is unsafe for some reason